File: python/isbn-verifier/isbn_verifier.py
import re


def is_valid(isbn):
    if not re.compile(r"^\d-?\d{3}-?\d{5}-?[\dX]$").match(isbn):
        return False

    sum = 0
    for i, c in enumerate(isbn.replace("-", "")[::-1]):
        if c == 'X':
            sum += 10
        else:
            sum += int(c) * (i + 1)

    return sum % 11 == 0

File: fsharp/isbn-verifier/IsbnVerifier.fs
module IsbnVerifier

open System.Text.RegularExpressions

let private hasValidCharacters isbn = Regex.IsMatch(isbn, @"^\d{9}[\dX]$")

let private digitToValue index digit =
    let indexFactor = index + 1
    match digit with
    | 'X' -> 10 * indexFactor
    | _ -> (int digit - int '0') * indexFactor

let private checksum isbn =
    isbn
    |> Seq.rev
    |> Seq.mapi digitToValue
    |> Seq.sum

let isValid (isbn: string) =
    let normalizedIsbn = isbn.Replace("-", "")
    hasValidCharacters normalizedIsbn && checksum normalizedIsbn % 11 = 0
